src/dialer/queue.js

```diff
@@ -0,0 +1,253 @@
+const once = require('once')
+const debug = require('debug')
+const log = debug('libp2p:switch:dial')
+log.error = debug('libp2p:switch:dial:error')
+
+const noop = () => {}
+
+/**
+ * Components required to execute a dial
+ * @typedef {Object} DialRequest
+ * @property {PeerInfo} peerInfo - The peer to dial to
+ * @property {string} [protocol] - The protocol to create a stream for
+ * @property {boolean} useFSM - If `callback` should return a ConnectionFSM
+ * @property {function(Error, Connection|ConnectionFSM)} callback
+ */
+
+/**
+ * A convenience array wrapper for controlling
+ * a per peer queue
+ *
+ * @returns {Queue}
+ */
+function Queue () {
+  let queue = []
+  let isRunning = false
+
+  return {
+    push: function (item) {
+      queue.push(item)
+    },
+    shift: function () {
+      return queue.shift()
+    },
+    isRunning: function () {
+      return isRunning
+    },
+    size: function () {
+      return queue.length
+    },
+    start: function () {
+      isRunning = true
+    },
+    stop: function () {
+      isRunning = false
+    }
+  }
+}
+
+class DialerQueue {
+  /**
+   * @constructor
+   * @param {Switch} _switch
+   */
+  constructor (_switch) {
+    this._queue = {}
+    this.switch = _switch
+  }
+
+  /**
+   * Iterates over all items in the DialerQueue
+   * and executes there callback with an error.
+   *
+   * This causes the entire DialerQueue to be drained
+   */
+  abort () {
+    const queues = Object.values(this._queue)
+    queues.forEach(queue => {
+      while (queue.size() > 0) {
+        let dial = queue.shift()
+        dial.callback(DIAL_ABORTED())
+      }
+    })
+  }
+
+  /**
+   * Adds the `dialRequest` to the queue and ensures the queue is running
+   *
+   * @param {DialRequest} dialRequest
+   */
+  add ({ peerInfo, protocol, useFSM, callback }) {
+    const id = peerInfo.id.toB58String()
+    const proxyConnection = new Connection()
+    proxyConnection.setPeerInfo(peerInfo)
+
+    callback = once(callback || noop)
+
+    let queue = this._queue[id] = this._queue[id] || new Queue()
+    queue.push({ protocol, proxyConnection, useFSM, callback })
+
+    if (!queue.isRunning()) {
+      log('starting dial queue to %s', id)
+      queue.start()
+      this._run(peerInfo)
+    }
+  }
+
+  /**
+   * Attempts to find a muxed connection for the given peer. If one
+   * isn't found, a new one will be created.
+   *
+   * Returns an array containing two items. The ConnectionFSM and wether
+   * or not the ConnectionFSM was just created. The latter can be used
+   * to determine dialing needs.
+   *
+   * @private
+   * @param {PeerInfo} peerInfo
+   * @returns {[ConnectionFSM, Boolean]}
+   */
+  _getOrCreateConnection (peerInfo) {
+    const id = peerInfo.id.toB58String()
+    let connectionFSM = this.switch.connection.getOne(id)
+    let didCreate = false
+
+    if (!connectionFSM) {
+      connectionFSM = new ConnectionFSM({
+        _switch: this.switch,
```

Comment:
This is due to switch being a reserved word. In other places incorrect spellings are used, which I am not a fan of as I it tends to lead to typo errors. The `_` is confusing though, as it can also indicate an unused parameter.  I think we can look at this in a subsequent PR to clean that up everywhere.